WIP: Search books

TatianaPan opened the pull request:
Now user can search books by author. 
- list of all books is rendered on at home page
- user can enter author last name o first name in search field and get all books of this author

There are 2 problems that I would like to discuss:
- now author's first and last name are stored in same column `author` -> list of books is ordered by author's first name, which is not very common I think. But I'm not sure if I should store author's name in separate columns (first name and last name, but if author has 3 names? `J.K.Rowling` for example. And I'll have to prompt user to fill 2 fields instead of one when adding a new book).

- usual search query `Book.where("author ILIKE ?", "%#{@author}%")` doesn't work case-insensitive for records in Russian. I found a working solution (`Book.where('author collate "en_US" ILIKE ?', "%#{@author}%")`), but not sure it's the best one/the right one. As far as I understand, I should either drop DB and recreate it and explicitly set different type of collation. Or try to set collation type on separate columns of DB. But it also seems like LC_COLLATE is dependent on operation system locale, so I'm not sure if it works for production: this `Book.where('author collate "ru_RU" ILIKE ?', "%#{@author}%")` didn't pass Semaphore pipeline.

As I mentioned, I would like to discuss it.
![Screenshot 2020-05-19 at 14 22 10](https://user-images.githubusercontent.com/48619229/82325961-5e2bb900-99dc-11ea-9eea-ed7714b35794.png)

![Screenshot 2020-05-19 at 14 22 21](https://user-images.githubusercontent.com/48619229/82325971-671c8a80-99dc-11ea-9ddd-09cf5646316f.png)

TatianaPan committed:
Add search field, action and tests

TatianaPan committed:
Fixed tests after merge with master

TatianaPan committed:
Add index on author column in books table

TatianaPan commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Search books](https://trello.com/c/MdPwpy43/6-search-books)

ahorincar commented on app/models/book.rb:
```diff
@@ -18,10 +18,14 @@ class Book < ApplicationRecord
 
+  # rubocop: disable Metrics/AbcSize
   def strip_input_fields
+    self.author = author.strip unless author.nil?
```
Maybe we can make this a bit more reusable by defining the columns to be stripped somewhere in this class, and then looping through those and stripping them. Something like:

```ruby
STRIPPABLE_ATTRIBUTES = %i[author title isbn10 isbn13]

def strip_input_fields
  STRIPPABLE_ATTRIBUTES.each { |attribute_name| send(attribute_name).strip }
end
```

The `strip_input_fields` could then go into a concern e.g. `Strippable`, which you can include in whichever model you may have strippable fields for (in this case the `Book` model):
```ruby
class Book
  include Strippable

STRIPPABLE_ATTRIBUTES = %i[author title isbn10 isbn13]

before_validation :strip_input_fields
```

TatianaPan replied:
We solved it during our pair programming session

ahorincar commented on app/views/home/_form.html.haml:
```diff
@@ -0,0 +1,7 @@
+= simple_form_for :search, url: root_path, method: "GET", html: { class: 'form-inline'} do |f|
+
+  .form-group
+    = f.input :author, placeholder: 'Author', input_html: { value: "" }, label: false
```
Is the `input_html: { value: "" }` part necessary?

TatianaPan replied:
Changed `f.input` to `f.input_field` and `input_html: { value: ' ' }` to `value: ' '`

ahorincar commented on app/views/home/index.html.haml:
```diff
@@ -1,2 +1,20 @@
+
+  = render 'form'
+
+  .mt-5.border.rounded-lg.shadow-sm
```
This looks very similar to the table from `books/index.html.haml`. Could we maybe extract this in a partial, which we use both in `books/index` and `home/index`?

TatianaPan replied:
Since these tables are a bit different, I decided not to extract them to partial yet.

ahorincar commented on test/system/home_test.rb:
```diff
@@ -0,0 +1,27 @@
+    end
+  end
+
+  test 'user searches a book by author in Russian' do
```
I would suggest adding a couple more tests, testing edge cases as well such as searching for an empty string, searching for a term which also has leading/trailing whitespaces, searching for a term which should not match anything, etc.

TatianaPan replied:
Added more tests

ahorincar commented on app/controllers/home_controller.rb:
```diff
@@ -1,7 +1,13 @@
+    return @all_books = Book.all.order(:author) if @search.blank?
+
+    @author = @search[:author].strip
+    @all_books = Book.where('author collate "en_US" ILIKE ?', "%#{@author}%")
```
What happens if instead of collating, you use the `lower` SQL function when performing the query? e.g. `Book.where('LOWER(author) LIKE ?', author.downcase)`

TatianaPan replied:
This is solved now with creating new database with explicitly set locale to `ru_RU.UTF-8`

ahorincar commented on app/controllers/home_controller.rb:
```diff
@@ -1,7 +1,13 @@
-  def index; end
+  def index
+    @search = params[:search]
+    return @all_books = Book.all.order(:author) if @search.blank?
```
What do you think about renaming `all_books` to just `books`? `all_books` implies it's going to hold every book in the DB, and we know that will not be the case when a search filter is applied.

TatianaPan replied:
That's a good point, thanks! I changed it.

ahorincar reviewed (commented):
This is a breaking change, but I would suggest using two columns for storing the author's first and last names. To answer your question regarding authors with multiple names: if the author has more than one family name e.g. `Laakman-McDowel`, this can be stored in the same column `last_name` for example. Same with first names, `J. K.` would be considered first names, so they could go in the same column `first_name`, whereas `Rowling` would go in `last_name`.

Also, is there a reason why the user can only search a book by its author, and not also by name?

TatianaPan committed:
Change collation on author and title in books table

TatianaPan committed:
Revert migration for collation set up

TatianaPan committed:
Recreated database with new collation and ctype

TatianaPan committed:
Add Dockerfile and change semaphore config to re-build postgres image

TatianaPan committed:
Add and rename columns on books table

TatianaPan committed:
Changed author to first_name and last_name

TatianaPan committed:
Implemented search with pg_search gem

TatianaPan committed:
Add default value to last_name

TatianaPan committed:
Add book decorator and book status icon to homepage

TatianaPan committed:
Add Strippable concern and test

TatianaPan committed:
Refactored search method and added more tests for it

TatianaPan commented:
@ahorincar  I added one more column on `Books` and split `author` into `first_name` and `last_name`. Problem with search in Russian has been solved by changing database configuration in `database.yml` and creating database with appropriate `locale`.

ahorincar commented on db/migrate/20200603113832_add_last_name_to_books.rb:
```diff
@@ -0,0 +1,9 @@
+class AddLastNameToBooks < ActiveRecord::Migration[6.0]
+  def change
+    add_column :books, :last_name, :string, null: false, default: ''
```
I think it might make sense at this point to introduce a new entity `Author`. It is a bit strange to have `first_name` and `last_name` on a book. And I'd say this is not just a matter of renaming these two columns. Since these two fields are clearly describing the author, it would make sense for them to live in a separate table `authors`. This will also simplify things if you end up adding even more info about the author (or more author-related logic).

ahorincar replied:
So ideally, there would be a new model `Authors`, as well as the `authors` table of course. I don't think you need any views/controllers for this entity, but you will need to add the association between a book and an author.

ahorincar replied:
I think the scenario is pretty similar to the one described here: https://guides.rubyonrails.org/kindle/association_basics.html

However, this addition does require some refactoring. It's not much, but still maybe you want to do this separate from this PR and then update this PR to work with `authors`.

ahorincar commented on test/models/concerns/strippable_test.rb:
```diff
@@ -0,0 +1,27 @@
+  test 'should strip trailing whitespaces on input fields' do
+    class ExampleClass
+      include Strippable
+      attr_accessor :username, :email, :phone
```
I like that you used a new class to test this module, rather than make use of `Book` :) Makes the test suite more reliable I think 👍

ahorincar commented on app/models/book.rb:
```diff
@@ -14,15 +19,14 @@ class Book < ApplicationRecord
-  def strip_input_fields
-    self.isbn10 = isbn10.strip unless isbn10.nil?
-    self.isbn13 = isbn13.strip unless isbn13.nil?
+  def decorate
```
Hmm, the model should not really know about any decorator-related logic. A standard way of adding decorator-related logic in Rails is to implement something similar to the view helpers. I recommend checking out [this](https://github.com/drapergem/draper) gem, it simplifies the process of building and accessing decorated objects. You can of course, also build this functionality from scratch but Draper simplifies this process.

ahorincar reviewed (commented):
Kudos for the extensive test suite that you added for the search! And also for the investigation into database collations!

I left a suggestion about improving the system design by adding a new entity for authors. I think that will make the code a bit more flexible to future changes.

TatianaPan committed:
Create author table and add reference to books

TatianaPan committed:
Remove author first and last name from books

TatianaPan committed:
Create Author model and validation test, add association to Book model

TatianaPan committed:
Add and changed fixtures

TatianaPan committed:
Refactored book and author models, controllers and view to add new books and authors

TatianaPan committed:
Removed book decorator and test

TatianaPan committed:
Fixed book controller and model tests (except of test for search)

TatianaPan committed:
Add tests for Author

TatianaPan committed:
Modified search and added tests

TatianaPan committed:
Delete book decorator

TatianaPan commented:
I added `Author` entity and I would propose to merge it within this PR (and finally close it:)).
I also removed decorator related stuff and will add it in a separate PR, as proposed.

TatianaPan committed:
Added author_id to permitted attributes to update author records correct

TatianaPan committed:
Merge branch 'master' into search-books
